File: ios/demo-krdpass-auth/AuthViewModel.swift
import Foundation
import SwiftUI
import KrdpassAuth

@MainActor
@Observable
final class AuthViewModel {
    // MARK: - State

    var isSigningIn = false

    var isLoadingUserInfo = false

    var isBusy = false

    var tokens: KrdpassTokenResult? {
        // Decode once per token set. The claim views read these on every redraw.
        didSet {
            idTokenClaims = Self.decodedClaims(tokens?.idToken, auth: krdpassAuth)
            accessTokenClaims = Self.decodedClaims(tokens?.accessToken, auth: krdpassAuth)
        }
    }

    /// Decoded ID / access token claims (display only, NOT signature-verified).
    private(set) var idTokenClaims: [String: JSONValue] = [:]
    private(set) var accessTokenClaims: [String: JSONValue] = [:]

    private static func decodedClaims(_ token: String?, auth: KrdpassAuth) -> [String: JSONValue] {
        guard let token else { return [:] }
        // Access tokens are often opaque, so a failed decode is normal: show nothing.
        return (try? auth.decodeTokenUnverified(token)) ?? [:]
    }

    var userInfo: KrdpassUserInfo?
    var errorMessage: String?

    /// App Store listing for KRDPASS, set only when sign-in failed with
    /// `provider_not_installed`. It is the one auth failure the user can fix, so the UI
    /// offers it as a button instead of leaving them at a dead end.
    var installUrl: URL?

    var includeCitizenScope = true
    var includeOfflineScope = true
    var useServerMode = true

    // MARK: - Dependencies

    /// Not private: the app's root view attaches `.withKrdpassDeepLinkHandling(auth)` to it.
    let krdpassAuth: KrdpassAuth
    private let backendService: AuthBackendService

    // MARK: - Init

    init() {
        let config = KrdpassConfig(
            clientId: Config.clientId,
            redirectUri: Config.redirectUri,
            environment: Config.environment
        )
        self.krdpassAuth = KrdpassAuth(config: config)
        self.backendService = AuthBackendService(baseUrl: Config.backendUrl)
    }

    // MARK: - Auth Config

    var currentConfig: KrdpassConfig {
        krdpassAuth.currentConfig
    }

    // MARK: - Sign In

    func signIn() async {
        guard !isSigningIn else { return }
        isSigningIn = true
        errorMessage = nil
        installUrl = nil

        do {
            var scopes = [KrdpassScopes.openid, KrdpassScopes.profile]
            if includeCitizenScope {
                scopes.append(KrdpassScopes.citizenIdentity)
            }
            if includeOfflineScope {
                scopes.append(KrdpassScopes.offlineAccess)
            }

            if useServerMode {
                try await signInWithServer(scopes: scopes)
            } else {
                try await signIn(scopes: scopes)
            }
        } catch let error as KrdpassError where error.code == "cancelled" {
            // The user backed out on purpose: no error. Matching on the code
            // also catches a Deny tap, which comes back from CAS on the
            // redirect as `authenticationFailed` with `code == "cancelled"`.
        } catch KrdpassError.timeout {
            errorMessage = "KRDPASS did not respond in time. Try signing in again."
        } catch KrdpassError.busy {
            errorMessage = "A sign-in is already in progress. Finish or cancel it first."
        } catch KrdpassError.providerNotInstalled(installUrl: let url) {
            // The only failure with a recovery action: the button does the work.
            errorMessage = "KRDPASS is not installed on this device."
            installUrl = url.flatMap(URL.init(string:))
        } catch {
            errorMessage = error.localizedDescription
        }

        isSigningIn = false
    }

    private func signIn(scopes: [String]) async throws {
        let result = try await krdpassAuth.signIn(scopes: scopes)
        self.tokens = result
    }

    /// Server mode: the backend runs PAR and token exchange, while the provider
    /// launch and result still go through the SDK.
    private func signInWithServer(scopes: [String]) async throws {
        let pkce = try krdpassAuth.generatePkcePair()
        let state = try krdpassAuth.generateState()
        let nonce = try krdpassAuth.generateState()

        let parResponse = try await backendService.getRequestUri(
            codeChallenge: pkce.codeChallenge,
            state: state,
            nonce: nonce,
            environment: Config.environment,
            redirectUri: Config.redirectUri,
            scope: scopes.joined(separator: " ")
        )

        // Don't wait past the request_uri the backend just minted; the backend
        // copies expires_in from CAS unvalidated, so floor it at 1s.
        let authTimeout = TimeInterval(max(1, parResponse.expiresIn ?? 300))
        let authResult = await krdpassAuth.authenticate(
            requestUri: parResponse.requestUri,
            state: parResponse.state ?? state,
            timeout: authTimeout
        )

        // AuthResult is a closed enum, so this switch is exhaustive: add a case to the SDK
        // and this stops compiling instead of silently landing in a `default`.
        switch authResult {
        case .success(let response):
            let tokenResponse = try await backendService.exchangeToken(
                code: response.code,
                state: parResponse.state ?? state,
                codeVerifier: pkce.codeVerifier
            )
            self.tokens = try Self.tokenResult(from: tokenResponse)

        // Map onto the same typed errors the direct signIn() path throws, so signIn() above
        // handles one set of outcomes whichever mode ran.
        case .cancelled:
            throw KrdpassError.userCancelled
        case .timeout:
            throw KrdpassError.timeout
        case .busy:
            throw KrdpassError.busy
        case .error(let authError):
            if authError.error == "provider_not_installed" {
                throw KrdpassError.providerNotInstalled(installUrl: authError.installUrl)
            }
            throw KrdpassError.authenticationFailed(authError.message, code: authError.error)
        }
    }

    /// `KrdpassTokenResult(dictionary:)` returns nil when `accessToken` is absent or empty,
    /// which means the backend returned no usable token. Fail rather than fabricate one.
    private static func tokenResult(from response: TokenResponseDTO) throws -> KrdpassTokenResult {
        guard let result = KrdpassTokenResult(dictionary: response.asDictionary) else {
            throw SignInError(message: "The backend returned no access token.")
        }
        return result
    }

    // MARK: - Logout

    func logout() async {
        let session = tokens
        clearSession()
        guard let session else { return }
        // Best effort: the local session is already gone and there is no screen left to
        // retry from, so never block sign-out on the network.
        try? await revokeSessionTokens(session)
    }

    /// Drop local state only. Callers that also need the grant revoked use `logout()`.
    private func clearSession() {
        tokens = nil
        userInfo = nil
        errorMessage = nil
        installUrl = nil
        isLoadingUserInfo = false
    }

    /// Revoke the refresh token first, then the access token. Order matters:
    /// the refresh token is what mints new access tokens, and revoking only the
    /// access token leaves the grant alive server-side.
    private func revokeSessionTokens(_ session: KrdpassTokenResult) async throws {
        var targets: [(String, String)] = []
        if let refreshToken = session.refreshToken { targets.append((refreshToken, "refresh_token")) }
        // An empty token would 400 at the server's own validation.
        if !session.accessToken.isEmpty { targets.append((session.accessToken, "access_token")) }

        for (token, hint) in targets {
            if useServerMode {
                try await backendService.revokeToken(
                    token: token,
                    environment: Config.environment,
                    tokenTypeHint: hint
                )
            } else {
                try await krdpassAuth.revokeToken(token: token, tokenTypeHint: hint)
            }
        }
    }

    // MARK: - User Info

    func fetchUserInfo() async {
        guard tokens != nil, !isLoadingUserInfo else { return }

        isLoadingUserInfo = true

        do {
            // Goes through validAccessToken() rather than `tokens?.accessToken`,
            // so expiry is handled where the token is used.
            let accessToken = try await validAccessToken()
            let info = try await krdpassAuth.getUserInfo(accessToken: accessToken)
            self.userInfo = info
            flashMessage(.ok("User info synced"))
            self.isLoadingUserInfo = false
        } catch {
            flashMessage(.failed("Sync failed: \(error.localizedDescription)"))
            self.isLoadingUserInfo = false
        }
    }

    /// The access token to send with the next API call, refreshed first if it
    /// has expired. Expiry is handled here, where the token is used, not by a
    /// button the user has to remember to press; `isExpired()` allows clock skew.
    private func validAccessToken() async throws -> String {
        guard let current = tokens else {
            throw SignInError(message: "Not signed in.")
        }
        guard current.isExpired() else { return current.accessToken }
        guard let refreshToken = current.refreshToken else {
            // No offline_access scope, so there is nothing to refresh with. Sign in again.
            throw SignInError(message: "Session expired. Sign in again.")
        }
        let refreshed = try await refreshedTokens(using: refreshToken, scope: current.scope)
        self.tokens = refreshed
        return refreshed.accessToken
    }

    /// Exchange a refresh token for a new token set, via the backend or the SDK.
    /// The scopes granted to the session are re-sent: an omitted `scope` leaves
    /// the decision to the server, which may silently narrow the grant.
    private func refreshedTokens(
        using refreshToken: String,
        scope: String?
    ) async throws -> KrdpassTokenResult {
        guard useServerMode else {
            return try await krdpassAuth.refreshTokens(refreshToken: refreshToken, scope: scope)
        }
        let response = try await backendService.refreshToken(
            refreshToken: refreshToken,
            environment: Config.environment,
            scope: scope
        )
        return try Self.tokenResult(from: response)
    }

    func clearError() {
        errorMessage = nil
        installUrl = nil
    }

    /// Merged claims from UserInfo (preferred) or ID Token
    private var claims: [String: JSONValue] {
        userInfo?.raw ?? idTokenClaims
    }

    private func claim(_ key: String) -> String? {
        claims[key]?.stringValue
    }

    var firstName: String {
        userInfo?.citizenFirst ?? claim("citizen_first") ?? ""
    }

    var fullName: String {
        userInfo?.citizenFullName ?? {
            let parts = [
                claim("citizen_first"),
                claim("citizen_second"),
                claim("citizen_third"),
                claim("citizen_surname")
            ].compactMap { $0 }.filter { !$0.isEmpty }

            if parts.isEmpty {
                return claim("upn") ?? "Citizen User"
            }
            return parts.joined(separator: " ")
        }()
    }

    var email: String {
        userInfo?.email ?? claim("email") ?? claim("upn") ?? "No email"
    }

    var birthdate: String? {
        userInfo?.birthdate ?? claim("birthdate")
    }

    var sex: String? {
        userInfo?.sexAtBirth ?? claim("sex_at_birth")
    }

    var profilePicUrl: String? {
        userInfo?.picture ?? claim("citizen_profile_picture")
    }

    // MARK: - Token Management Actions

    var actionMessage: ActionMessage?
    private var actionMessageToken = 0

    /// Show a transient status message that auto-clears after 3s. A monotonic token guards the
    /// clear, so a newer message is never wiped early by an earlier call's timer.
    private func flashMessage(_ message: ActionMessage) {
        actionMessage = message
        actionMessageToken += 1
        let token = actionMessageToken
        Task {
            try? await Task.sleep(nanoseconds: 3 * 1_000_000_000)
            if actionMessageToken == token { actionMessage = nil }
        }
    }

    func verifyToken() async {
        guard !isBusy else { return }
        guard let idToken = tokens?.idToken else {
            flashMessage(.failed("No ID token to verify"))
            return
        }
        isBusy = true
        actionMessage = nil
        do {
            // Verify signature using JWKS (audience is derived from the configured clientId)
            let _ = try await krdpassAuth.verifyToken(idToken: idToken)
            flashMessage(.ok("Token signature valid"))
        } catch {
            flashMessage(.failed("Invalid: \(error.localizedDescription)"))
        }
        isBusy = false
    }

    func refreshToken() async {
        guard !isBusy else { return }
        guard let current = tokens, let token = current.refreshToken else {
            flashMessage(.failed("No refresh token available"))
            return
        }
        isBusy = true
        actionMessage = nil
        do {
            self.tokens = try await refreshedTokens(using: token, scope: current.scope)
            flashMessage(.ok("Tokens refreshed"))
        } catch {
            flashMessage(.failed("Refresh failed: \(error.localizedDescription)"))
        }
        isBusy = false
    }

    func revokeToken() async {
        guard !isBusy else { return }
        guard let session = tokens else {
            flashMessage(.failed("No token to revoke"))
            return
        }
        isBusy = true
        actionMessage = nil
        do {
            try await revokeSessionTokens(session)
            clearSession()
            flashMessage(.ok("Tokens revoked, signed out"))
        } catch {
            flashMessage(.failed("Revoke failed: \(error.localizedDescription)"))
        }
        isBusy = false
    }
}

// MARK: - Action Message

struct ActionMessage {
    let ok: Bool
    let text: String

    static func ok(_ text: String) -> ActionMessage { ActionMessage(ok: true, text: text) }
    static func failed(_ text: String) -> ActionMessage { ActionMessage(ok: false, text: text) }
}

// MARK: - Errors

/// Lightweight error that carries the SDK's canonical message for display.
struct SignInError: LocalizedError {
    let message: String
    var errorDescription: String? { message }
}

File: flutter/lib/auth_view_model.dart
import 'dart:async';

import 'package:demo_krdpass_auth/config.dart';
import 'package:demo_krdpass_auth/models/action_message.dart';
import 'package:demo_krdpass_auth/services/auth_backend_service.dart';
import 'package:flutter/foundation.dart';
import 'package:krdpass_auth_flutter/krdpass_auth_flutter.dart';

String _describe(Object e) => switch (e) {
  KrdpassException(:final message) => message,
  BackendException(:final message) => message,
  _ => e.toString(),
};

/// Owns the demo's auth session and SDK orchestration, off the widget tree, so the
/// views stay a function of this state and the session survives a rebuild.
class AuthViewModel extends ChangeNotifier {
  /// The backend is a dependency, not a global: same shape as the Android and iOS
  /// samples, and it is the seam a test would replace.
  final AuthBackendService _backend = AuthBackendService(
    baseUrl: Config.backendUrl,
  );

  KrdpassTokenResult? _tokens;
  KrdpassUserInfo? _userInfo;
  String? _error;
  String? _installUrl;
  ActionMessage? _actionMessage;
  Timer? _actionMessageTimer;

  bool _signingIn = false;
  bool _loadingUserInfo = false;
  bool _busy = false;

  bool _includeCitizenScope = true;
  bool _includeOfflineScope = true;
  bool _useServerMode = true;

  bool get isLoggedIn => _tokens != null;
  KrdpassTokenResult? get tokens => _tokens;
  KrdpassUserInfo? get userInfo => _userInfo;
  String? get error => _error;

  /// Store listing for the KRDPASS app, set only when sign-in failed with
  /// `provider_not_installed`. It is the one auth failure the user can fix, so the
  /// UI offers it as a link instead of leaving them at a dead end.
  String? get installUrl => _installUrl;
  ActionMessage? get actionMessage => _actionMessage;

  /// A sign-in is in flight.
  bool get isSigningIn => _signingIn;

  /// A UserInfo sync is in flight.
  bool get isLoadingUserInfo => _loadingUserInfo;

  /// A token-management action (verify / refresh / revoke) is in flight.
  bool get isBusy => _busy;

  bool get includeCitizenScope => _includeCitizenScope;
  bool get includeOfflineScope => _includeOfflineScope;
  bool get useServerMode => _useServerMode;

  set includeCitizenScope(bool value) {
    _includeCitizenScope = value;
    notifyListeners();
  }

  set includeOfflineScope(bool value) {
    _includeOfflineScope = value;
    notifyListeners();
  }

  set useServerMode(bool value) {
    _useServerMode = value;
    notifyListeners();
  }

  @override
  void dispose() {
    _actionMessageTimer?.cancel();
    super.dispose();
  }

  /// Sign in, handling each outcome on its own terms: a cancellation is not a
  /// failure, a timeout is retryable, and `provider_not_installed` carries the
  /// install URL that fixes it.
  Future<void> signIn() async {
    if (_signingIn) return;
    _signingIn = true;
    _error = null;
    _installUrl = null;
    notifyListeners();
    try {
      _tokens = _useServerMode
          ? await _signInWithServer()
          : await _signInDirect();
    } on KrdpassCancelledException {
      // The user backed out on purpose: drop the spinner, show no error.
    } on KrdpassTimeoutException {
      _error = 'KRDPASS did not respond in time. Try signing in again.';
    } on KrdpassBusyException {
      _error = 'A sign-in is already in progress. Finish or cancel it first.';
    } on KrdpassAuthenticationException catch (e) {
      // installUrl is non-null only for provider_not_installed. Passing it through is
      // what turns "something went wrong" into a working next step.
      _error = e.message;
      _installUrl = e.installUrl;
    } catch (e) {
      _error = _describe(e);
    } finally {
      _signingIn = false;
      notifyListeners();
    }
  }

  Future<KrdpassTokenResult> _signInWithServer() async {
    final pkce = KrdpassAuth.instance.generatePkcePair();
    final state = KrdpassAuth.instance.generateState();
    final nonce = KrdpassAuth.instance.generateState();

    final parResponse = await _backend.getRequestUri(
      codeChallenge: pkce.codeChallenge,
      state: state,
      nonce: nonce,
      environment: Config.environment,
      redirectUri: Config.redirectUri,
      scope: _scopes.join(' '),
    );

    // Don't wait past the request_uri the backend just minted; the backend copies
    // expires_in from CAS unvalidated, so floor it at 1s.
    final authTimeoutSeconds = parResponse.expiresIn ?? 300;
    final result = await KrdpassAuth.instance.authenticate(
      requestUri: parResponse.requestUri,
      state: parResponse.state,
      timeout: Duration(
        seconds: authTimeoutSeconds < 1 ? 1 : authTimeoutSeconds,
      ),
    );

    if (!result.isSuccess) {
      // Map onto the same exception hierarchy the direct signIn() throws, so
      // signIn() above handles one set of outcomes whichever mode ran.
      if (result.isCancelled) throw const KrdpassCancelledException();
      if (result.isTimeout) throw const KrdpassTimeoutException();
      if (result.isBusy) throw const KrdpassBusyException();
      throw KrdpassAuthenticationException(
        result.errorMessage,
        code: result.error,
        installUrl: result.installUrl,
      );
    }

    return _backend.exchangeToken(
      code: result.code!,
      // The backend echoes the state it bound to the request; fall back to the one we
      // generated rather than to '', which the server would reject.
      state: parResponse.state ?? state,
      codeVerifier: pkce.codeVerifier,
    );
  }

  Future<KrdpassTokenResult> _signInDirect() => KrdpassAuth.instance.signIn(
    scopes: _scopes,
    timeout: const Duration(minutes: 5),
  );

  List<String> get _scopes => [
    KrdpassScopes.openid,
    KrdpassScopes.profile,
    if (_includeCitizenScope) KrdpassScopes.citizenIdentity,
    if (_includeOfflineScope) KrdpassScopes.offlineAccess,
  ];

  Future<void> fetchUserInfo() async {
    if (_loadingUserInfo || _tokens == null) return;
    _loadingUserInfo = true;
    notifyListeners();
    try {
      // Goes through _validAccessToken() rather than tokens.accessToken, so
      // expiry is handled where the token is used.
      _userInfo = await KrdpassAuth.instance.getUserInfo(
        accessToken: await _validAccessToken(),
      );
      _showStatus(const ActionMessage.ok('User info synced'));
    } catch (e) {
      _showStatus(ActionMessage.failed('Sync failed: ${_describe(e)}'));
    } finally {
      _loadingUserInfo = false;
      notifyListeners();
    }
  }

  /// The access token to send with the next API call, refreshed first if it has
  /// expired. Expiry is handled here, where the token is used, not by a button
  /// the user has to remember to press; `isExpired` allows for clock skew.
  Future<String> _validAccessToken() async {
    final current = _tokens;
    if (current == null) throw StateError('Not signed in.');
    if (!current.isExpired()) return current.accessToken;
    final refreshToken = current.refreshToken;
    // No offline_access scope means there is nothing to refresh with: sign in again.
    if (refreshToken == null) {
      throw StateError('Session expired. Sign in again.');
    }
    final refreshed = await _refreshedTokens(current, refreshToken);
    _tokens = refreshed;
    return refreshed.accessToken;
  }

  /// Exchange a refresh token for a new token set, via the backend or the SDK.
  /// The scopes granted to the session are re-sent: an omitted `scope` leaves
  /// the decision to the server, which may silently narrow the grant.
  Future<KrdpassTokenResult> _refreshedTokens(
    KrdpassTokenResult current,
    String refreshToken,
  ) {
    if (_useServerMode) {
      return _backend.refreshToken(
        refreshToken: refreshToken,
        environment: Config.environment,
        scope: current.scope,
      );
    }
    return KrdpassAuth.instance.refreshTokens(
      refreshToken: refreshToken,
      scope: current.scope,
    );
  }

  /// Refresh on demand, so the demo can show the exchange happening. Real code
  /// should not need this button: [_validAccessToken] refreshes at the point of use.
  Future<void> refreshToken() async {
    if (_busy) return;
    final current = _tokens;
    final refreshToken = current?.refreshToken;
    if (current == null || refreshToken == null) {
      _showStatus(const ActionMessage.failed('No refresh token available'));
      return;
    }
    _busy = true;
    notifyListeners();
    try {
      _tokens = await _refreshedTokens(current, refreshToken);
      _showStatus(const ActionMessage.ok('Tokens refreshed'));
    } catch (e) {
      _showStatus(ActionMessage.failed('Refresh failed: ${_describe(e)}'));
    } finally {
      _busy = false;
      notifyListeners();
    }
  }

  /// Revoke the session's tokens. The refresh token is the one that matters: it is the
  /// long-lived credential, and revoking it is what ends the grant.
  Future<void> revokeToken() async {
    if (_busy) return;
    final current = _tokens;
    if (current == null) {
      _showStatus(const ActionMessage.failed('No token to revoke'));
      return;
    }
    _busy = true;
    notifyListeners();
    try {
      await _revokeSessionTokens(current, _useServerMode);
      _clearSession();
      _showStatus(const ActionMessage.ok('Tokens revoked, signed out'));
    } catch (e) {
      _showStatus(ActionMessage.failed('Revoke failed: ${_describe(e)}'));
    } finally {
      _busy = false;
      notifyListeners();
    }
  }

  Future<void> verifyToken() async {
    if (_busy) return;
    final idToken = _tokens?.idToken;
    if (idToken == null) {
      _showStatus(const ActionMessage.failed('No ID token to verify'));
      return;
    }
    _busy = true;
    notifyListeners();
    try {
      await KrdpassAuth.instance.verifyToken(idToken: idToken);
      _showStatus(const ActionMessage.ok('Token signature valid'));
    } catch (e) {
      _showStatus(ActionMessage.failed('Invalid: ${_describe(e)}'));
    } finally {
      _busy = false;
      notifyListeners();
    }
  }

  /// Revoke the refresh token first, then the access token. Order matters: the
  /// refresh token is what mints new access tokens, and revoking only the
  /// access token leaves the grant alive server-side.
  Future<void> _revokeSessionTokens(
    KrdpassTokenResult session,
    bool useServerMode,
  ) async {
    final targets = <(String, String)>[
      if (session.refreshToken != null) (session.refreshToken!, 'refresh_token'),
      if (session.accessToken.isNotEmpty) (session.accessToken, 'access_token'),
    ];
    for (final (token, hint) in targets) {
      if (useServerMode) {
        await _backend.revokeToken(
          token: token,
          environment: Config.environment,
          tokenTypeHint: hint,
        );
      } else {
        await KrdpassAuth.instance.revokeToken(
          token: token,
          tokenTypeHint: hint,
        );
      }
    }
  }

  /// Sign out. Clearing the local fields is the visible half; the half that
  /// matters is revoking the refresh token, which would otherwise keep working.
  /// There is no end-session endpoint here, so issued access tokens stay valid
  /// until they expire; if your deployment adds one, call it here as well.
  Future<void> logout() async {
    final session = _tokens;
    final useServerMode = _useServerMode;
    _clearSession();
    if (session == null) return;
    try {
      await _revokeSessionTokens(session, useServerMode);
    } catch (_) {
      // Best effort: the local session is already gone and there is no screen left to
      // retry from, so never block sign-out on the network.
    }
  }

  void clearError() {
    _error = null;
    _installUrl = null;
    notifyListeners();
  }

  void _clearSession() {
    _tokens = null;
    _userInfo = null;
    _error = null;
    _installUrl = null;
    _loadingUserInfo = false;
    _actionMessageTimer?.cancel();
    _actionMessage = null;
    notifyListeners();
  }

  /// Show a transient status line that clears itself after 3s. The timer is replaced on
  /// every message, so a newer one is never wiped early by an earlier call's timer.
  void _showStatus(ActionMessage message) {
    _actionMessage = message;
    _actionMessageTimer?.cancel();
    _actionMessageTimer = Timer(const Duration(seconds: 3), () {
      _actionMessage = null;
      notifyListeners();
    });
    notifyListeners();
  }
}
